In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
df = pd.read_csv('data/mob_money_preprocessed_2.csv')

In [3]:
df1 = df.copy()

In [4]:
df1.head()

transaction_reference payment_reference    merchant    customer_name  \
0   KPY-CA-5xlpa6ZxSxHf         923438867  BLENET LTD  Emmanuel Norman   
1   KPY-CA-bVBQDQuNVhsb         923446863  BLENET LTD    Edward Gyamfi   
2   KPY-CA-YtBlORY4dc0S         923448600  BLENET LTD       Elom Ohene   
3   KPY-CA-s93p6xaZ3TOJ         923449039  BLENET LTD   Benneth Pascal   
4   KPY-CA-UjVz7bbxlKV0         923449546  BLENET LTD   Benneth Pascal   

                customer_email  customer_phone_number  merchant_bears_cost  \
0        iamnormkay@icloud.com           2.335435e+11                    1   
1     gyamfiedward98@gmail.com           2.335575e+11                    1   
2  humanitarianalarm@gmail.com           2.332457e+11                    1   
3           mrpubenr@gmail.com           2.335582e+11                    1   
4           mrpubenr@gmail.com           2.335582e+11                    1   

             tran_date   payment_created_at   payment_updated_at  ...  \
0  2024-08-26 16:03:35  2024-08-26 16:03:35  2024-08-26 16:15:17  ...   
1  2024-08-26 16:17:07  2024-08-26 16:17:07  2024-08-26 16:17:41  ...   
2  2024-08-26 16:20:13  2024-08-26 16:20:13  2024-08-26 16:21:04  ...   
3  2024-08-26 16:20:50  2024-08-26 16:20:50  2024-08-26 16:37:42  ...   
4  2024-08-26 16:21:46  2024-08-26 16:21:46  2024-08-26 16:22:39  ...   

   payment_reversals_type currency  auth_model mobile_number_for_mobile_money  \
0                       0      GHS  STK_PROMPT                   233543521075   
1                       0      GHS  STK_PROMPT                   233557469987   
2                       0      GHS  STK_PROMPT                   233245689134   
3                       0      GHS  STK_PROMPT                   233272525000   
4                       0      GHS  STK_PROMPT                   233243721306   

  mobile_network_provider         processor_response  \
0                     MTN        Transaction failed.   
1                     MTN    Transaction successful.   
2                     MTN    Transaction successful.   
3                  Airtel  Transaction id is invalid   
4                     MTN    Transaction successful.   

        kora_translated_processor_response  amount amount_collected is_fraud  
0  Charge attempt failed, please try again  468.05             0.00        0  
1                        Charge successful  156.02           156.02        0  
2                        Charge successful  156.02           156.02        0  
3  Charge attempt failed, please try again  506.00             0.00        0  
4                        Charge successful  506.00           506.00        0  

[5 rows x 27 columns]

In [5]:
df1['customer_phone_number'] = df1['customer_phone_number'].astype(str)

In [11]:
df1.shape

(396845, 27)

## EDA

In [13]:
fraud_df = df1[df1['is_fraud']==1]
normal_df = df1[df1['is_fraud']==0]

In [15]:
fraud_df.is_fraud.value_counts()

is_fraud
1    143322
Name: count, dtype: int64

#### merchant

### customer_name

In [17]:
pd.set_option('display.max_columns', 500)
pd.set_option('display.max_rows', 500)
fraud_df.customer_name.value_counts().head(40)

customer_name
Ibrahim Issaka              19331
Albert Baidoo               14262
Enock Gyabi                  2941
Francis Francis              2220
James Hackman                1557
Efua Gyaawa                  1477
Shadrack Ofosu Antwi         1377
Richmond Donkor              1006
Nana Qwaku                    969
Michael Nkansah               792
Because Because               783
Long Gee                      713
Emmanuel Nyarko               628
Cletus Adeina                 585
Alexander Addo                483
EmmanuelAnkomah               476
Mahmoud Razak                 445
Kweku Nkansah                 422
Baba Adama                    370
Know Opuku                    350
Top Shatta                    342
PrinceAgyiri                  331
YtfggvGgff                    331
Bock Money                    324
Richmond Dacosta Dacosta      316
Patrick Addo                  315
NANA Gaza                     303
Isaac Affum                   295
Prosper Ziem                  290


In [19]:
unique_customer_names = fraud_df.customer_name.unique()
for name in unique_customer_names:
    print(name)

Edmond Asiedu
Gyinae Josephine
Samuel Karikari
Benneth Pascal
Esther Tiwaa
Kofi Adu
Seth Yawson
Khadija Nasiru
Joshua Lebibu Adamu
Eleke Okoronkwo
Shaiba Abubakar
Michael Addy
Boateng Derrick Frimpong
Agbo Emma
Richmond Odardze
Amidu Abdul Hakim
Kwasi Tchato
Joyce Owusu
Collins Antwi
Collins Fosu
Fosu Collins
Baba Rashida
Grace Addo
Bernard Botchway
Patrick Mensah
Baba Zakiatu
Martha Boateng
AKOSUA SIKA
Sakyibea Adjei
Emmanuel Nana Kwesi Amanquah
Nathaniel Acorlor
RICHARD SIA
Samata Mohammed
Felix Arthur
Elijah Tagoe
Emmanuella Nketiah
Markins Boadu
Nyarko Theophlius
Dickson Jeffrey
Charity Tetteh
SIA RICHARD
Ibrahim Iddrisu
Abubakar Sadick
Victoria  Agyei
Alhassan Mariam
Yy Yuy
-
David Dumfeh
Darkwa Charles
Charles Darkwa
Gilbert Akoto
Oracle Funds
Abelumbolga Louis
Ruth Darko
Daniel Akuffo
Osei David
Michael Adzraku
Duke Edem Kemeh
Nicholas Mensah
Daniel Fosu Akuffo
Lawrence Gyan
Roland Kendeh
Akugyinu Daniel
Fauzia Salifu Abubakar
Mohammed Kuti
Sarah Adams
Landry Rudasingwa
Cosmos A

In [21]:
unique_customer_emails = fraud_df.customer_email.unique()
for email in unique_customer_emails:
    print(email)

edmondasiedu6@gmail.com
gyinaejosephine@gmail.com
annorblay8@gmail.com
mrpubenr@gmail.com
francisobeng2468@gmail.com
adudavid042@gmail.com
sethyawson2019@gmail.com
nasirukhadija49@gmail.com
joshualebibuadamu9611@gmail.com
pacoeleke@gmail.com
shaibaabubakar0@gmail.com
aryeegrace1987@gmail.com
boatengderrickfrimpong413@gmail.com
eagbo1213@gmail.com
rodardze@gmail.com
nuhuhardi5@gmail.com
tsathokossididier1@gmail.com
joyceokofo597@gmail.com
freakylord1718@gmail.com
1kolley2002@gmail.com
babazakiatu@gmail.com
addo95186@gmail.com
bernardbotchway1594@gmail.com
shalpason@yahoo.com
marthaboateng2847@gmail.com
bernardbotchway2914@gmail.com
akosuasika6119@gmail.com
sakyibeaadjei9628@gmail.com
bernardbotchway7200@gmail.com
pearltin92@gmail.com
nathanielacorlor3424@gmail.com
siarichard741@gmail.com
faroukabdulff111@gmail.com
felixarthur6549@gmail.com
elijahtagoe4051@gmail.com
emmanuellanketiah6239@gmail.com
markinsboadu@gmail.com
theophliusn199@gmail.com
dicksonbentum@gmail.com
loveadjoa238@gmail.

In [23]:
df3 = pd.concat([normal_df, fraud_df], axis=0)

In [25]:
df3.is_fraud.value_counts()

is_fraud
0    253523
1    143322
Name: count, dtype: int64

In [27]:
df3.head()

transaction_reference payment_reference    merchant    customer_name  \
0   KPY-CA-5xlpa6ZxSxHf         923438867  BLENET LTD  Emmanuel Norman   
1   KPY-CA-bVBQDQuNVhsb         923446863  BLENET LTD    Edward Gyamfi   
2   KPY-CA-YtBlORY4dc0S         923448600  BLENET LTD       Elom Ohene   
3   KPY-CA-s93p6xaZ3TOJ         923449039  BLENET LTD   Benneth Pascal   
4   KPY-CA-UjVz7bbxlKV0         923449546  BLENET LTD   Benneth Pascal   

                customer_email customer_phone_number  merchant_bears_cost  \
0        iamnormkay@icloud.com        233543521075.0                    1   
1     gyamfiedward98@gmail.com        233557469987.0                    1   
2  humanitarianalarm@gmail.com        233245689134.0                    1   
3           mrpubenr@gmail.com        233558163994.0                    1   
4           mrpubenr@gmail.com        233558163994.0                    1   

             tran_date   payment_created_at   payment_updated_at  \
0  2024-08-26 16:03:35  2024-08-26 16:03:35  2024-08-26 16:15:17   
1  2024-08-26 16:17:07  2024-08-26 16:17:07  2024-08-26 16:17:41   
2  2024-08-26 16:20:13  2024-08-26 16:20:13  2024-08-26 16:21:04   
3  2024-08-26 16:20:50  2024-08-26 16:20:50  2024-08-26 16:37:42   
4  2024-08-26 16:21:46  2024-08-26 16:21:46  2024-08-26 16:22:39   

   for_disbursement_wallet account_created_date  account_risk_level  \
0                        0  2023-04-26 10:54:42  above_average_risk   
1                        0  2023-04-26 10:54:42  above_average_risk   
2                        0  2023-04-26 10:54:42  above_average_risk   
3                        0  2023-04-26 10:54:42  above_average_risk   
4                        0  2023-04-26 10:54:42  above_average_risk   

  payment_source_type processor payment_status payment_channel  \
0        mobile_money     nsano         failed             api   
1        mobile_money     nsano        success             api   
2        mobile_money     nsano        success             api   
3        mobile_money     nsano         failed             api   
4        mobile_money     nsano        success             api   

   payment_reversals_type currency  auth_model  \
0                       0      GHS  STK_PROMPT   
1                       0      GHS  STK_PROMPT   
2                       0      GHS  STK_PROMPT   
3                       0      GHS  STK_PROMPT   
4                       0      GHS  STK_PROMPT   

   mobile_number_for_mobile_money mobile_network_provider  \
0                    233543521075                     MTN   
1                    233557469987                     MTN   
2                    233245689134                     MTN   
3                    233272525000                  Airtel   
4                    233243721306                     MTN   

          processor_response       kora_translated_processor_response  amount  \
0        Transaction failed.  Charge attempt failed, please try again  468.05   
1    Transaction successful.                        Charge successful  156.02   
2    Transaction successful.                        Charge successful  156.02   
3  Transaction id is invalid  Charge attempt failed, please try again  506.00   
4    Transaction successful.                        Charge successful  506.00   

   amount_collected  is_fraud  
0              0.00         0  
1            156.02         0  
2            156.02         0  
3              0.00         0  
4            506.00         0

In [29]:
df1 = df3.copy()

In [31]:
df1.drop(columns=['for_disbursement_wallet'], axis=1, inplace=True)

In [142]:
df1.account_risk_level.value_counts()

account_risk_level
above_average_risk    237457
high_risk             159388
Name: count, dtype: int64

#### payment_source_type

In [33]:
df1.drop(columns=['payment_source_type'], axis=1, inplace=True)

#### processor

In [35]:
df1.drop(columns=['processor'], axis=1, inplace=True)

In [436]:
df1.payment_channel.value_counts()

payment_channel
api      323077
modal     73768
Name: count, dtype: int64

In [438]:
fraud_df.payment_channel.value_counts()

payment_channel
api      87322
modal    56000
Name: count, dtype: int64

In [440]:
normal_df.payment_channel.value_counts()

payment_channel
api      235755
modal     17768
Name: count, dtype: int64

In [166]:
df1.payment_reversals_type.value_counts()

payment_reversals_type
0    396843
1         2
Name: count, dtype: int64

#### currency

In [37]:
df1.drop(columns=['currency'], axis=1, inplace=True)

#### auth_model

In [39]:
df1.drop(columns=['auth_model'], axis=1, inplace=True)

In [179]:
df1.mobile_network_provider.value_counts()

mobile_network_provider
MTN         347262
Vodafone     47618
Airtel        1965
Name: count, dtype: int64

In [181]:
fraud_df.mobile_network_provider.value_counts()

mobile_network_provider
MTN    143322
Name: count, dtype: int64

 - all fraud txns are MTN

In [185]:
df1.processor_response.value_counts()

processor_response
Transaction successful.                                                                                                                        154059
TARGET_AUTHORIZATION_ERROR                                                                                                                     108705
Transaction failed.                                                                                                                             41073
SUCCESS                                                                                                                                         30286
AUTHORIZATION_SENDER_ACCOUNT_NOT_ACTIVE                                                                                                         24403
                                                                                                                                                ...  
You have paid GHS300.00 to 18010001 in reference to KORAPAY. Your new balance is 

In [187]:
fraud_df.processor_response.value_counts()

processor_response
TARGET_AUTHORIZATION_ERROR                 108705
AUTHORIZATION_SENDER_ACCOUNT_NOT_ACTIVE     24403
ACCOUNTHOLDER_WITH_FRI_NOT_FOUND             6700
RESOURCE_NOT_FOUND                           3514
Name: count, dtype: int64

In [190]:
df1.kora_translated_processor_response.value_counts()

kora_translated_processor_response
Charge successful                                                                                                             185785
Insufficient funds or transaction limit reached                                                                                88010
Charge attempt failed, please try again                                                                                        46385
We are unable to complete this transaction at this time, please try again later                                                31291
Your wallet is not active. Please contact your financial institution                                                           22059
PIN request timed out                                                                                                           6801
The recipient wallet is invalid                                                                                                 6099
An error occurred, please try agai

In [192]:
fraud_df.kora_translated_processor_response.value_counts()

kora_translated_processor_response
Insufficient funds or transaction limit reached                                                      88008
Charge attempt failed, please try again                                                              24127
Your wallet is not active. Please contact your financial institution                                 22059
The recipient wallet is invalid                                                                       6099
We are unable to complete this transaction at this time, please try again later                       2889
An error occurred, please try again                                                                    137
You will receive a prompt on mobile number. Kindly enter your wallet PIN to authorize the payment        3
Name: count, dtype: int64

- Insufficient funds or transaction limit reached: 88008/88010 (suspicious data/total data)
- Charge attempt failed, please try again: 24127/46385 (suspicious data/total data)
- Your wallet is not active. Please contact your financial institution: 22059/22059 (suspicious data/total data)
- The recipient wallet is invalid: 6099/6099 (suspicious data/total data)
- We are unable to complete this transaction at this time, please try again later: 2889/31291 (suspicious data/total data)
- An error occurred, please try again: 137/5920 (suspicious data/total data)
- You will receive a prompt on mobile number. Kindly enter your wallet PIN to authorize the payment: 3/3 (suspicious data/total data)

In [194]:
df1.columns

Index(['transaction_reference', 'payment_reference', 'merchant',
       'customer_name', 'customer_email', 'customer_phone_number',
       'merchant_bears_cost', 'tran_date', 'payment_created_at',
       'payment_updated_at', 'account_created_date', 'account_risk_level',
       'payment_status', 'payment_channel', 'payment_reversals_type',
       'mobile_number_for_mobile_money', 'mobile_network_provider',
       'processor_response', 'kora_translated_processor_response', 'amount',
       'amount_collected', 'is_fraud'],
      dtype='object')

#### temporal analysis

In [41]:
# ensuring both variables are in datetime format
df1['payment_created_at'] = pd.to_datetime(df1['payment_created_at'])
df1['payment_updated_at'] = pd.to_datetime(df1['payment_updated_at'])

# extracting features from 'payment_created_on'
df1['created_dow'] = df1['payment_created_at'].dt.day_name()  # dow (day of week the txn was created)
df1['created_hod'] = df1['payment_created_at'].dt.hour  # hod (hour of day the txn was created i.e 0-23)
df1['created_month'] = df1['payment_created_at'].dt.month_name()  # month txn was created

# extracting features from 'payment_completed_on'
df1['completed_dow'] = df1['payment_updated_at'].dt.day_name()
df1['completed_hod'] = df1['payment_updated_at'].dt.hour
df1['completed_month'] = df1['payment_updated_at'].dt.month_name()

In [43]:
df1['payment_completion_time_secs'] = (df1['payment_updated_at'] - df1['payment_created_at']).dt.total_seconds()

In [45]:
# ANALYSING TIME-BASED PATTERNS
# Count transactions by day of week for fraud and normal
fraud_by_day = df1[df1['is_fraud'] == 1]['created_dow'].value_counts()
normal_by_day = df1[df1['is_fraud'] == 0]['created_dow'].value_counts()

# Count transactions by hour for fraud and normal
fraud_by_hour = df1[df1['is_fraud'] == 1]['created_hod'].value_counts()
normal_by_hour = df1[df1['is_fraud'] == 0]['created_hod'].value_counts()

# Count transactions by month for fraud and normal
fraud_by_month = df1[df1['is_fraud'] == 1]['created_month'].value_counts()
normal_by_month = df1[df1['is_fraud'] == 0]['created_month'].value_counts()

most of the fraud txns are executed in December

#### looking for insights in normal txns executed by suspicious customers

In [47]:
sus_customer_txns_names = df1[df1['customer_name'].isin(unique_customer_names)]
sus_customer_txns_emails = df1[df1['customer_email'].isin(unique_customer_emails)]
sus_customer_all_txns = pd.concat([sus_customer_txns_names, sus_customer_txns_emails])
sus_customer_all_txns = sus_customer_all_txns.drop_duplicates() 

sus_customer_all_txns.head()

transaction_reference payment_reference    merchant   customer_name  \
3    KPY-CA-s93p6xaZ3TOJ         923449039  BLENET LTD  Benneth Pascal   
4    KPY-CA-UjVz7bbxlKV0         923449546  BLENET LTD  Benneth Pascal   
14   KPY-CA-kWuhUZyyzLPA         923478449  BLENET LTD    Billy Kowuvi   
16   KPY-CA-QN5oLU26Szev         923480055  BLENET LTD    Billy Kowuvi   
18   KPY-CA-0G1NwC9Qzgw4         923483678  BLENET LTD    Billy Kowuvi   

           customer_email customer_phone_number  merchant_bears_cost  \
3      mrpubenr@gmail.com        233558163994.0                    1   
4      mrpubenr@gmail.com        233558163994.0                    1   
14  billyhanz06@gmail.com        233559245228.0                    1   
16  billyhanz06@gmail.com        233559245228.0                    1   
18  billyhanz06@gmail.com        233559245228.0                    1   

              tran_date  payment_created_at  payment_updated_at  \
3   2024-08-26 16:20:50 2024-08-26 16:20:50 2024-08-26 16:37:42   
4   2024-08-26 16:21:46 2024-08-26 16:21:46 2024-08-26 16:22:39   
14  2024-08-26 17:13:17 2024-08-26 17:13:17 2024-08-26 17:25:25   
16  2024-08-26 17:16:18 2024-08-26 17:16:18 2024-08-26 17:28:17   
18  2024-08-26 17:23:03 2024-08-26 17:23:03 2024-08-26 17:35:17   

   account_created_date  account_risk_level payment_status payment_channel  \
3   2023-04-26 10:54:42  above_average_risk         failed             api   
4   2023-04-26 10:54:42  above_average_risk        success             api   
14  2023-04-26 10:54:42  above_average_risk         failed             api   
16  2023-04-26 10:54:42  above_average_risk         failed             api   
18  2023-04-26 10:54:42  above_average_risk         failed             api   

    payment_reversals_type  mobile_number_for_mobile_money  \
3                        0                    233272525000   
4                        0                    233243721306   
14                       0                    233559245228   
16                       0                    233559245228   
18                       0                    233559245228   

   mobile_network_provider         processor_response  \
3                   Airtel  Transaction id is invalid   
4                      MTN    Transaction successful.   
14                     MTN        Transaction failed.   
16                     MTN        Transaction failed.   
18                     MTN        Transaction failed.   

         kora_translated_processor_response  amount  amount_collected  \
3   Charge attempt failed, please try again   506.0               0.0   
4                         Charge successful   506.0             506.0   
14  Charge attempt failed, please try again   169.0               0.0   
16  Charge attempt failed, please try again   169.0               0.0   
18  Charge attempt failed, please try again   169.0               0.0   

    is_fraud created_dow  created_hod created_month completed_dow  \
3          0      Monday           16        August        Monday   
4          0      Monday           16        August        Monday   
14         0      Monday           17        August        Monday   
16         0      Monday           17        August        Monday   
18         0      Monday           17        August        Monday   

    completed_hod completed_month  payment_completion_time_secs  
3              16          August                        1012.0  
4              16          August                          53.0  
14             17          August                         728.0  
16             17          August                         719.0  
18             17          August                         734.0

In [49]:
sus_customer_all_txns.is_fraud.value_counts()

is_fraud
1    143322
0    101257
Name: count, dtype: int64

In [51]:
sus_customer_normal_txns = sus_customer_all_txns[sus_customer_all_txns['is_fraud']==0]
sus_customer_normal_txns.is_fraud.value_counts()

is_fraud
0    101257
Name: count, dtype: int64

In [53]:
sus_customer_sus_txns = sus_customer_all_txns[sus_customer_all_txns['is_fraud']==1]
print("Suspicious Transactions:")
print(sus_customer_sus_txns.describe())

Suspicious Transactions:
       merchant_bears_cost             payment_created_at  \
count        143322.000000                         143322   
mean              0.999860  2024-12-13 15:21:52.125947136   
min               0.000000            2024-08-26 15:57:03   
25%               1.000000  2024-12-10 06:04:05.249999872   
50%               1.000000            2024-12-17 02:25:44   
75%               1.000000            2024-12-28 19:05:03   
max               1.000000            2025-01-20 23:55:34   
std               0.011812                            NaN   

                  payment_updated_at  payment_reversals_type  \
count                         143322                143322.0   
mean   2024-12-14 05:17:33.775058688                     0.0   
min              2024-08-26 15:57:10                     0.0   
25%    2024-12-10 07:42:02.750000128                     0.0   
50%       2024-12-17 03:37:26.500000                     0.0   
75%              2024-12-28 21:10:19     

In [55]:
print("\nNormal Transactions (of suspicious customers):")
print(sus_customer_normal_txns.describe())


Normal Transactions (of suspicious customers):
       merchant_bears_cost             payment_created_at  \
count        101257.000000                         101257   
mean              0.999951  2024-12-06 10:29:42.970886400   
min               0.000000            2024-08-26 16:20:50   
25%               1.000000            2024-11-12 17:39:04   
50%               1.000000            2024-12-11 20:07:08   
75%               1.000000            2024-12-27 19:26:31   
max               1.000000            2025-01-20 23:58:50   
std               0.007027                            NaN   

                  payment_updated_at  payment_reversals_type  \
count                         101257                101257.0   
mean   2024-12-06 17:56:31.257473536                     0.0   
min              2024-08-26 16:22:39                     0.0   
25%              2024-11-13 12:39:35                     0.0   
50%              2024-12-11 20:09:42                     0.0   
75%              2

#### fraud_df statistics (count: 143322)
- amount: mean(496.38), min(10), 25%(200), 50%(400), 75%(500), max(5000), std(555.76)
- amount collected: mean(0), min(0), 25%(0), 50%(0), 75%(0), max(0), std(0)
- created time of day: mean(12.16), min(0), 25%(7), 50%(12), 75%(18), max(23), std(6.96)
- completed time of day: mean(12.12), min(0), 25%(7), 50%(12), 75%(18), max(23), std(7.01)
- payment_completion_time_secs: mean(50141.7), min(3), 25%(6), 50%(7), 75%(7480), max(3052648.0), std(376631.1)

#### normal_df (of sus customers) statistics (count: 101257)
- amount: mean(605.77), min(10), 25%(165), 50%(260), 75%(500), max(5000), std(1108.26)
- amount collected: mean(352.4), min(0), 25%(0), 50%(163), 75%(300), max(5000), std(732.25)
- created time of day: mean(12.61), min(0), 25%(9), 50%(13), 75%(17), max(23), std(5.82)
- completed time of day: mean(12.72), min(0), 25%(9), 50%(13), 75%(17), max(23), std(5.91)
- payment_completion_time_secs: mean(26808.3), min(0), 25%(31), 50%(36), 75%(426), max(3052648.0), std(278749.8)


In [57]:
# looking at stats of normal customers NOT found in sus data
norm_customer_txns_names = df1[~df1['customer_name'].isin(unique_customer_names)]
norm_customer_txns_emails = df1[~df1['customer_email'].isin(unique_customer_emails)]
norm_customer_all_txns = pd.concat([norm_customer_txns_names, norm_customer_txns_emails])
norm_customer_all_txns = norm_customer_all_txns.drop_duplicates()

In [58]:
# just checking...
norm_customer_all_txns.is_fraud.value_counts()

is_fraud
0    165683
Name: count, dtype: int64

In [61]:
print("\nNormal Transactions of non-suspicious customers:")
print(norm_customer_all_txns.describe())


Normal Transactions of non-suspicious customers:
       merchant_bears_cost             payment_created_at  \
count        165683.000000                         165683   
mean              0.999879  2024-11-26 17:46:57.080116736   
min               0.000000            2024-08-26 16:03:35   
25%               1.000000     2024-10-24 13:56:02.500000   
50%               1.000000            2024-11-26 18:16:19   
75%               1.000000            2024-12-23 09:59:23   
max               1.000000            2025-01-20 23:59:20   
std               0.010986                            NaN   

                  payment_updated_at  payment_reversals_type  \
count                         165683           165683.000000   
mean   2024-11-26 17:53:18.551275776                0.000012   
min              2024-08-26 16:15:17                0.000000   
25%       2024-10-24 13:58:36.500000                0.000000   
50%              2024-11-26 18:19:56                0.000000   
75%       2024-1

### feature engineering

In [63]:
%%time

# calculating last digit difference
def calculate_phone_number_diff(group):
    group['mobile_number_for_mobile_money'] = pd.to_numeric(group['mobile_number_for_mobile_money'], errors='coerce')
    group['phone_number_diff'] = group['mobile_number_for_mobile_money'].diff()
    group['phone_number_diff'] = group['phone_number_diff'].fillna(0)  # Or -999 for first transaction
    return group

df1 = df1.groupby('customer_email').apply(calculate_phone_number_diff)

CPU times: user 16min 13s, sys: 11.3 s, total: 16min 24s
Wall time: 16min 33s


<timed exec>:8: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.


In [65]:
df1.head()

transaction_reference payment_reference  \
customer_email                                                              
001daman100@gmail.com      131119   KPY-CA-r0hEfAhffbv0    38262010121656   
003kfx@gmail.com           52770    KPY-CA-BVbXPVkbZ10N         217666212   
                           92356    KPY-CA-BD6zJ3HWAdAy         224006476   
                           139152   KPY-CA-6j8jNhqtKdh5         230551425   
004.teyeemmanuel@gmail.com 139042   KPY-CA-9Q6fbJPs3fC0    27635010124566   

                                           merchant         customer_name  \
customer_email                                                              
001daman100@gmail.com      131119  Nexasoft Limited        David Amankwah   
003kfx@gmail.com           52770             Exness         EVANS KOOMSON   
                           92356             Exness         EVANS KOOMSON   
                           139152            Exness         EVANS KOOMSON   
004.teyeemmanuel@gmail.com 139042  Nexasoft Limited  Samuel  Teye Amartey   

                                               customer_email  \
customer_email                                                  
001daman100@gmail.com      131119       001daman100@gmail.com   
003kfx@gmail.com           52770             003kfx@gmail.com   
                           92356             003kfx@gmail.com   
                           139152            003kfx@gmail.com   
004.teyeemmanuel@gmail.com 139042  004.teyeemmanuel@gmail.com   

                                  customer_phone_number  merchant_bears_cost  \
customer_email                                                                 
001daman100@gmail.com      131119        233508671529.0                    1   
003kfx@gmail.com           52770         233550144312.0                    1   
                           92356         233550144312.0                    1   
                           139152        233550144312.0                    1   
004.teyeemmanuel@gmail.com 139042        233540227425.0                    1   

                                             tran_date  payment_created_at  \
customer_email                                                               
001daman100@gmail.com      131119  2024-12-05 02:50:34 2024-12-05 02:50:34   
003kfx@gmail.com           52770   2024-10-25 14:36:25 2024-10-25 14:36:25   
                           92356   2024-11-14 18:51:27 2024-11-14 18:51:27   
                           139152  2024-12-08 16:27:46 2024-12-08 16:27:46   
004.teyeemmanuel@gmail.com 139042  2024-12-08 15:00:54 2024-12-08 15:00:54   

                                   payment_updated_at account_created_date  \
customer_email                                                               
001daman100@gmail.com      131119 2024-12-05 02:51:05  2023-07-27 12:36:23   
003kfx@gmail.com           52770  2024-10-25 14:37:04  2021-04-21 11:55:06   
                           92356  2024-11-14 18:52:05  2021-04-21 11:55:06   
                           139152 2024-12-08 16:28:16  2021-04-21 11:55:06   
004.teyeemmanuel@gmail.com 139042 2024-12-08 15:01:26  2023-07-27 12:36:23   

                                   account_risk_level payment_status  \
customer_email                                                         
001daman100@gmail.com      131119  above_average_risk        success   
003kfx@gmail.com           52770            high_risk        success   
                           92356            high_risk        success   
                           139152           high_risk        success   
004.teyeemmanuel@gmail.com 139042  above_average_risk        success   

                                  payment_channel  payment_reversals_type  \
customer_email                                                              
001daman100@gmail.com      131119             api                       0   
003kfx@gmail.com           52770              api                       0   
                           923

In [67]:
# AGGREGATING CUSTOMER METRICS
# - Number of unique phone numbers: how many different phone numbers each customer uses. obviously, a high count is suspicious.
# - Number of txns with different numbers: how many txns involve a change in phone number from the previous txn
# - Proportion of txns with different numbers: dividing the number of txns with different numbers by the total number of txns for the customer
# - Average/median difference: calculating the average or median absolute difference between phone numbers (excluding 0 differences, 
#                              which are legit same-number entries). Small average differences are highly suspicious.

# calculating metrics and merging directly into df1
def calculate_customer_metrics(group):
    unique_phone_numbers = group['mobile_number_for_mobile_money'].nunique()
    diff_number_txns = (group['phone_number_diff']!= 0).sum()
    total_transactions = len(group)
    proportion_different = diff_number_txns / total_transactions if total_transactions > 0 else 0
    differences = group['phone_number_diff'].abs()
    differences = differences[differences!= 0]
    average_difference = differences.mean() if not differences.empty else 0

    return pd.Series({
        'unique_phone_numbers': unique_phone_numbers,
        'diff_number_txns': diff_number_txns,
        'total_transactions': total_transactions,
        'proportion_different_numbers': proportion_different,
        'average_phone_number_difference': average_difference
    })

In [69]:
df2 = df1.copy()
df2 = df2.rename(columns={'customer_email': 'cust_email'}) # renaming customer_email to avoid errors
df2 = df2.reset_index()
df2 = df2.join(df2.groupby('customer_email').apply(calculate_customer_metrics), on='customer_email', how='left')

/var/folders/66/4xtpccc53fbdkdpdddgj1rxc0000gn/T/ipykernel_79489/1848200746.py:4: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df2 = df2.join(df2.groupby('customer_email').apply(calculate_customer_metrics), on='customer_email', how='left')


In [71]:
df2.head()

customer_email  level_1 transaction_reference  \
0       001daman100@gmail.com   131119   KPY-CA-r0hEfAhffbv0   
1            003kfx@gmail.com    52770   KPY-CA-BVbXPVkbZ10N   
2            003kfx@gmail.com    92356   KPY-CA-BD6zJ3HWAdAy   
3            003kfx@gmail.com   139152   KPY-CA-6j8jNhqtKdh5   
4  004.teyeemmanuel@gmail.com   139042   KPY-CA-9Q6fbJPs3fC0   

  payment_reference          merchant         customer_name  \
0    38262010121656  Nexasoft Limited        David Amankwah   
1         217666212            Exness         EVANS KOOMSON   
2         224006476            Exness         EVANS KOOMSON   
3         230551425            Exness         EVANS KOOMSON   
4    27635010124566  Nexasoft Limited  Samuel  Teye Amartey   

                   cust_email customer_phone_number  merchant_bears_cost  \
0       001daman100@gmail.com        233508671529.0                    1   
1            003kfx@gmail.com        233550144312.0                    1   
2            003kfx@gmail.com        233550144312.0                    1   
3            003kfx@gmail.com        233550144312.0                    1   
4  004.teyeemmanuel@gmail.com        233540227425.0                    1   

             tran_date  payment_created_at  payment_updated_at  \
0  2024-12-05 02:50:34 2024-12-05 02:50:34 2024-12-05 02:51:05   
1  2024-10-25 14:36:25 2024-10-25 14:36:25 2024-10-25 14:37:04   
2  2024-11-14 18:51:27 2024-11-14 18:51:27 2024-11-14 18:52:05   
3  2024-12-08 16:27:46 2024-12-08 16:27:46 2024-12-08 16:28:16   
4  2024-12-08 15:00:54 2024-12-08 15:00:54 2024-12-08 15:01:26   

  account_created_date  account_risk_level payment_status payment_channel  \
0  2023-07-27 12:36:23  above_average_risk        success             api   
1  2021-04-21 11:55:06           high_risk        success             api   
2  2021-04-21 11:55:06           high_risk        success             api   
3  2021-04-21 11:55:06           high_risk        success             api   
4  2023-07-27 12:36:23  above_average_risk        success             api   

   payment_reversals_type  mobile_number_for_mobile_money  \
0                       0                    233592381050   
1                       0                    233550144312   
2                       0                    233550144312   
3                       0                    233550144312   
4                       0                    233540227425   

  mobile_network_provider       processor_response  \
0                     MTN  Transaction successful.   
1                     MTN  Transaction successful.   
2                     MTN  Transaction successful.   
3                     MTN  Transaction successful.   
4                     MTN  Transaction successful.   

  kora_translated_processor_response  amount  amount_collected  is_fraud  \
0                  Charge successful   153.0             153.0         0   
1                  Charge successful  1000.0            1000.0         0   
2                  Charge successful   170.0             170.0         0   
3                  Charge successful   165.0             165.0         0   
4                  Charge successful   149.0             149.0         0   

  created_dow  created_hod created_month completed_dow  completed_hod  \
0    Thursday            2      December      Thursday              2   
1      Friday           14       October        Friday             14   
2    Thursday           18      November      Thursday             18   
3      Sunday           16      December        Sunday             16   
4      Sunday           15      December        Sunday             15   

  completed_month  payment_completion_time_secs  phone_number_diff  \
0        December                          31.0                0.0   
1         October                          39.0                0.0   
2        November                          38.0                0.0   
3        December                          30.0                0.0

In [73]:
# Identify Shared Phone Numbers with Different Customer Info (Binary Flag)
shared_phone_numbers = df2.groupby('mobile_number_for_mobile_money', as_index=False).filter(lambda x: len(x) > 1)
shared_phone_numbers = shared_phone_numbers.groupby('mobile_number_for_mobile_money', as_index=False).agg({'customer_email': 'nunique'}) # Count unique customer_ids for each phone
shared_phone_numbers = shared_phone_numbers[shared_phone_numbers['customer_email'] > 1] #Keep phone numbers shared by more than one customer
shared_phone_numbers['shared_phone_number_flag'] = 1  # Create the binary flag
shared_phone_numbers = shared_phone_numbers[['mobile_number_for_mobile_money', 'shared_phone_number_flag']] #Keep only the phone number and the shared_phone_number_flag

df2 = pd.merge(df2, shared_phone_numbers, on='mobile_number_for_mobile_money', how='left')
df2['shared_phone_number_flag'] = df2['shared_phone_number_flag'].fillna(0)

In [75]:
df2.head()

customer_email  level_1 transaction_reference  \
0       001daman100@gmail.com   131119   KPY-CA-r0hEfAhffbv0   
1            003kfx@gmail.com    52770   KPY-CA-BVbXPVkbZ10N   
2            003kfx@gmail.com    92356   KPY-CA-BD6zJ3HWAdAy   
3            003kfx@gmail.com   139152   KPY-CA-6j8jNhqtKdh5   
4  004.teyeemmanuel@gmail.com   139042   KPY-CA-9Q6fbJPs3fC0   

  payment_reference          merchant         customer_name  \
0    38262010121656  Nexasoft Limited        David Amankwah   
1         217666212            Exness         EVANS KOOMSON   
2         224006476            Exness         EVANS KOOMSON   
3         230551425            Exness         EVANS KOOMSON   
4    27635010124566  Nexasoft Limited  Samuel  Teye Amartey   

                   cust_email customer_phone_number  merchant_bears_cost  \
0       001daman100@gmail.com        233508671529.0                    1   
1            003kfx@gmail.com        233550144312.0                    1   
2            003kfx@gmail.com        233550144312.0                    1   
3            003kfx@gmail.com        233550144312.0                    1   
4  004.teyeemmanuel@gmail.com        233540227425.0                    1   

             tran_date  payment_created_at  payment_updated_at  \
0  2024-12-05 02:50:34 2024-12-05 02:50:34 2024-12-05 02:51:05   
1  2024-10-25 14:36:25 2024-10-25 14:36:25 2024-10-25 14:37:04   
2  2024-11-14 18:51:27 2024-11-14 18:51:27 2024-11-14 18:52:05   
3  2024-12-08 16:27:46 2024-12-08 16:27:46 2024-12-08 16:28:16   
4  2024-12-08 15:00:54 2024-12-08 15:00:54 2024-12-08 15:01:26   

  account_created_date  account_risk_level payment_status payment_channel  \
0  2023-07-27 12:36:23  above_average_risk        success             api   
1  2021-04-21 11:55:06           high_risk        success             api   
2  2021-04-21 11:55:06           high_risk        success             api   
3  2021-04-21 11:55:06           high_risk        success             api   
4  2023-07-27 12:36:23  above_average_risk        success             api   

   payment_reversals_type  mobile_number_for_mobile_money  \
0                       0                    233592381050   
1                       0                    233550144312   
2                       0                    233550144312   
3                       0                    233550144312   
4                       0                    233540227425   

  mobile_network_provider       processor_response  \
0                     MTN  Transaction successful.   
1                     MTN  Transaction successful.   
2                     MTN  Transaction successful.   
3                     MTN  Transaction successful.   
4                     MTN  Transaction successful.   

  kora_translated_processor_response  amount  amount_collected  is_fraud  \
0                  Charge successful   153.0             153.0         0   
1                  Charge successful  1000.0            1000.0         0   
2                  Charge successful   170.0             170.0         0   
3                  Charge successful   165.0             165.0         0   
4                  Charge successful   149.0             149.0         0   

  created_dow  created_hod created_month completed_dow  completed_hod  \
0    Thursday            2      December      Thursday              2   
1      Friday           14       October        Friday             14   
2    Thursday           18      November      Thursday             18   
3      Sunday           16      December        Sunday             16   
4      Sunday           15      December        Sunday             15   

  completed_month  payment_completion_time_secs  phone_number_diff  \
0        December                          31.0                0.0   
1         October                          39.0                0.0   
2        November                          38.0                0.0   
3        December                          30.0                0.0

In [77]:
df2.customer_email.nunique()

87430

In [79]:
df2.shared_phone_number_flag.value_counts()

shared_phone_number_flag
0.0    351480
1.0     45365
Name: count, dtype: int64

In [81]:
df2['shared_phone_number_flag']= df2.shared_phone_number_flag.astype(int)

In [83]:
df2['unique_phone_numbers']= df2.unique_phone_numbers.astype(int)
df2['diff_number_txns']= df2.diff_number_txns.astype(int)
df2['total_transactions']= df2.total_transactions.astype(int)

In [85]:
# frequency of txns per customer
df2['txn_count_per_customer'] = df2.groupby('customer_email')['transaction_reference'].transform('count')

In [87]:
# frequency of txns per customer per hour
df2['txn_count_per_customer_per_hour'] = df2.groupby(['customer_email', 'created_hod'])['transaction_reference'].transform('count')

In [91]:
# maximum transactions within an hour (per day)
# df2.drop(columns=['max_txns_in_an_hour'], inplace=True)
max_txns_per_hour_per_customer = df2.groupby(['customer_email', 'created_dow', 'created_hod'])['transaction_reference'].count().groupby(['customer_email', 'created_dow']).max()
df2 = df2.merge(max_txns_per_hour_per_customer.rename('max_txns_in_an_hour_per_customer'), left_on=['customer_email', 'created_dow'], right_index=True, how='left')

In [93]:
# frequency of txns per customer per day
df2['txn_count_per_customer_per_day'] = df2.groupby(['customer_email', 'created_dow'])['transaction_reference'].transform('count')
df2.head()

customer_email  level_1 transaction_reference  \
0       001daman100@gmail.com   131119   KPY-CA-r0hEfAhffbv0   
1            003kfx@gmail.com    52770   KPY-CA-BVbXPVkbZ10N   
2            003kfx@gmail.com    92356   KPY-CA-BD6zJ3HWAdAy   
3            003kfx@gmail.com   139152   KPY-CA-6j8jNhqtKdh5   
4  004.teyeemmanuel@gmail.com   139042   KPY-CA-9Q6fbJPs3fC0   

  payment_reference          merchant         customer_name  \
0    38262010121656  Nexasoft Limited        David Amankwah   
1         217666212            Exness         EVANS KOOMSON   
2         224006476            Exness         EVANS KOOMSON   
3         230551425            Exness         EVANS KOOMSON   
4    27635010124566  Nexasoft Limited  Samuel  Teye Amartey   

                   cust_email customer_phone_number  merchant_bears_cost  \
0       001daman100@gmail.com        233508671529.0                    1   
1            003kfx@gmail.com        233550144312.0                    1   
2            003kfx@gmail.com        233550144312.0                    1   
3            003kfx@gmail.com        233550144312.0                    1   
4  004.teyeemmanuel@gmail.com        233540227425.0                    1   

             tran_date  payment_created_at  payment_updated_at  \
0  2024-12-05 02:50:34 2024-12-05 02:50:34 2024-12-05 02:51:05   
1  2024-10-25 14:36:25 2024-10-25 14:36:25 2024-10-25 14:37:04   
2  2024-11-14 18:51:27 2024-11-14 18:51:27 2024-11-14 18:52:05   
3  2024-12-08 16:27:46 2024-12-08 16:27:46 2024-12-08 16:28:16   
4  2024-12-08 15:00:54 2024-12-08 15:00:54 2024-12-08 15:01:26   

  account_created_date  account_risk_level payment_status payment_channel  \
0  2023-07-27 12:36:23  above_average_risk        success             api   
1  2021-04-21 11:55:06           high_risk        success             api   
2  2021-04-21 11:55:06           high_risk        success             api   
3  2021-04-21 11:55:06           high_risk        success             api   
4  2023-07-27 12:36:23  above_average_risk        success             api   

   payment_reversals_type  mobile_number_for_mobile_money  \
0                       0                    233592381050   
1                       0                    233550144312   
2                       0                    233550144312   
3                       0                    233550144312   
4                       0                    233540227425   

  mobile_network_provider       processor_response  \
0                     MTN  Transaction successful.   
1                     MTN  Transaction successful.   
2                     MTN  Transaction successful.   
3                     MTN  Transaction successful.   
4                     MTN  Transaction successful.   

  kora_translated_processor_response  amount  amount_collected  is_fraud  \
0                  Charge successful   153.0             153.0         0   
1                  Charge successful  1000.0            1000.0         0   
2                  Charge successful   170.0             170.0         0   
3                  Charge successful   165.0             165.0         0   
4                  Charge successful   149.0             149.0         0   

  created_dow  created_hod created_month completed_dow  completed_hod  \
0    Thursday            2      December      Thursday              2   
1      Friday           14       October        Friday             14   
2    Thursday           18      November      Thursday             18   
3      Sunday           16      December        Sunday             16   
4      Sunday           15      December        Sunday             15   

  completed_month  payment_completion_time_secs  phone_number_diff  \
0        December                          31.0                0.0   
1         October                          39.0                0.0   
2        November                          38.0                0.0   
3        December                          30.0                0.0

In [97]:
# checking out phone number pattern + high transaction amount
def check_phone_amount(group):
    group['phone_txn_amount_suspicious'] = 0
    for customer_email in group['customer_email'].unique():
        customer_group = group[group['customer_email'] == customer_email]
        phone_numbers = customer_group['mobile_number_for_mobile_money'].unique()
        if len(phone_numbers) > 1: #Check if the customer has more than one phone number
            high_amounts = customer_group[customer_group['amount'] > customer_group['amount'].quantile(0.90)] #Check for high transaction amounts
            if not high_amounts.empty: #If there are high transaction amounts, flag them
                group.loc[high_amounts.index, 'phone_txn_amount_suspicious'] = 1
    return group
df2 = df2.groupby('customer_email').apply(check_phone_amount).reset_index(drop=True)

/var/folders/66/4xtpccc53fbdkdpdddgj1rxc0000gn/T/ipykernel_79489/3523896357.py:12: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df2 = df2.groupby('customer_email').apply(check_phone_amount).reset_index(drop=True)


In [99]:
df2.phone_txn_amount_suspicious.value_counts()

phone_txn_amount_suspicious
0    387927
1      8918
Name: count, dtype: int64

In [101]:
# sudden spikes in txn count per customer
# Here, I'm using a rolling window (7 days) to calculate a moving average of txn counts per customer.
# The window size can be adjusted as needed. The spike flag is set if the current txn count is significantly 
# higher (I set it to 2.5x here) than the rolling average.
df2['txn_count_per_customer'] = df2.groupby('customer_email')['transaction_reference'].transform('count')
df2['rolling_txn_count_per_customer'] = df2.groupby('customer_email')['transaction_reference'].transform(lambda x: x.rolling(window=7, min_periods=1).count()) # 7-day rolling window
df2['txn_spike_flag'] = (df2['txn_count_per_customer'] > 2.5 * df2['rolling_txn_count_per_customer']).astype(int) # Example: 2.5x the rolling average

In [103]:
df2.txn_spike_flag.value_counts()

txn_spike_flag
1    218655
0    178190
Name: count, dtype: int64

In [105]:
# unusual txn amounts per customer
# flagging txns that are 5x higher than a customer's average txn for cases where a customer's total txn count is >=10
df2['avg_txn_amount_per_customer'] = df2.groupby('customer_email')['amount'].transform('mean')
# Identify transactions where total_transactions > 9
total_txn_threshold = df2['total_transactions'] > 9
df2['unusual_txn_flag'] = 0
df2.loc[total_txn_threshold, 'unusual_txn_flag'] = (df2.loc[total_txn_threshold, 'amount'] > 
                                                    5 * df2.loc[total_txn_threshold, 'avg_txn_amount_per_customer']).astype(int)

In [107]:
df2.unusual_txn_flag.value_counts()

unusual_txn_flag
0    395431
1      1414
Name: count, dtype: int64

In [109]:
df2.to_csv('data/mob_money_fe_dataset_v2.csv', index=False)

## Findings

- the merchant bears the cost of the transactions in 98.7% of the fraud instances.
- for normal txns, txn count is distributed evenly throughout the week but for fraud txns, 19.37% of all fraud instances were performed on Friday. Wednesdays and Tuesday have 16.2% and 14.6% respectively; Thursday is the lowest with 8.7%; the other days have similar distributions (for fraud txns).
- for hour of the day, a high number of fraud txns are performed by 7-8am (20) and 10-11am (24).
- for transaction count by month, 14.6% of the normal txns were made in October while 10.9% were made in September; the rest were distributed evenly among the months. however, 21% of the fraud txns were made in May while 11.5% were made in April; the rest were distributed evenly among the months.
- for merchants, Exness contributed to 73 out of 235 of the fraud txns (with a fraud rate of 0.000036%); Bet9ja (fraud rate of 0.000316%) and Betwinner (fraud rate of 0.000119%)(both betting companies) contributed to 33 and 24 of the 235 fraud txns respectively. Involsys (fraud rate of 0.000043%) contributed to 21 out of 235 of total fraud txns. Nexasoft strangely had a fraud rate of 100%, as 10 of the 10 txns in this dataset were all fraudulent.
- for payment method, while ~85% of the normal data are bank transfer txns, 10.14% are card txns, 2.9% are mobilemoney txns. 67.2% of the fraud txns performed were bank transfer txns, 18.2% are card txns, and 5.1% are mobilemoney txns.
- for payment_channel, 81% of normal txns are modal. the rest are API. ~90% of normal txns are modal. the rest are API.
- .
- all merchants associated with fraud are either high_risk or above_average_risk.
- wema processed most of the fraud txns (139/235). interswitch and sterling came in 2nd (37/235) and 3rd (29/235). jambopay had the processed the second lowest fraud txns (4/235) but had the highest fraud rate (0.001110%).
- only 7/235 fraudulent txns had a payment_status of expired, 5/235 had failed and 4/235 had rejected. the rest were successful.
- for currency, 91.5% of normal txns were NGN txns, 5.4% were USD txns, and 2.6% were KES txns. however, 87.75% of fraud txns were NGN txns and the rest were KES txns.
- out of 235 fraud txns, 28 were identified as anomalies (unusually high/low but in this case, all were unusually high) in terms of transaction amount so only 88.1% were in the 'normal' range of txn amount.
- for payment reversals, 9.8% of txns are reversals in fraud txns.
- the t-test showed a significant difference in payment completion times between fraud and normal transactions.

- the merchant bears the cost of the transactions in 98.7% of the fraud instances.
- for normal txns, txn count is distributed evenly throughout the week but for fraud txns, 19.37% of all fraud instances were performed on Friday. Wednesdays, Sunday and Friday have 16.2% and 14.6% respectively; Thursday is the lowest with 8.7%; the other days have similar distributions (for fraud txns).
- for hour of the day, a high number of fraud txns are performed by 7-8am (20) and 10-11am (24).
- for transaction count by month, 14.6% of the normal txns were made in October while 10.9% were made in September; the rest were distributed evenly among the months. however, 21% of the fraud txns were made in May while 11.5% were made in April; the rest were distributed evenly among the months.
- for merchants, merchantA contributed to 73 out of 235 of the fraud txns (with a fraud rate of 0.000036%); merchantB (fraud rate of 0.000316%) and merchantC (fraud rate of 0.000119%)(both betting companies) contributed to 33 and 24 of the 235 fraud txns respectively. merchantD (fraud rate of 0.000043%) contributed to 21 out of 235 of total fraud txns. merchantE strangely had a fraud rate of 100%, as 10 of the 10 txns in this dataset were all fraudulent.
- for payment method, while ~85% of the normal data are bank transfer txns, 10.14% are card txns, 2.9% are mobilemoney txns. 67.2% of the fraud txns performed were bank transfer txns, 18.2% are card txns, and 5.1% are mobilemoney txns.
for payment_channel, 81% of normal txns are modal. the rest are API. ~90% of normal txns are modal. the rest are API.
- all merchants associated with fraud are either high_risk or above_average_risk.
- processorA processed most of the fraud txns (139/235). processorB and processorC came in 2nd (37/235) and 3rd (29/235). processorD had the processed the second lowest fraud txns (4/235) but had the highest fraud rate (0.001110%).
only 7/235 fraudulent txns had a payment_status of expired, 5/235 had failed and 4/235 had rejected. the rest were successful.
- for currency, 91.5% of normal txns were NGN txns, 5.4% were USD txns, and 2.6% were KES txns. however, 87.75% of fraud txns were NGN txns and the rest were KES txns.
- out of 235 fraud txns, 28 were identified as anomalies (unusually high/low but in this case, all were unusually high) in terms of transaction amount so only 88.1% were in the 'normal' range of txn amount.
- for payment reversals, 9.8% of txns are reversals in fraud txns.
- the t-test showed a significant difference in payment completion times between fraud and normal transactions.

## Considerations for feature engineering

- drop merchant_bears_cost
- encode payment_description using a target encoding approach, where the encoding reflects the likelihood of fraud for each description.
- create a binary flag (is_suspicious_description) for transactions with these two suspicious payment descriptions.
- extract the day of the week as a categorical feature (txn_day_of_week) and one-hot encode it.
- add a feature (is_monday) to flag transactions that occur on Monday.
- Extract the hour of the day as a numerical feature (txn_hour).
- Add a binary flag (is_peak_hour) for transactions occurring at 10 AM.
- Extract the month as a categorical feature (txn_month) and one-hot encode it.
- Add a binary flag (is_december) for transactions in December.
- Use one-hot encoding for payment_source_type.
- Create a binary flag (is_bank_transfer) for bank transfer transactions.
- Encode the merchant as a categorical variable (merchant) with one-hot encoding or target encoding.
- Add a feature (merchant_risk_level) based on risk levels: high, above average, and low.
- Create binary flags for is_merchantA and is_merchantB.
- Encode the processor (processor) as a categorical feature with one-hot encoding.
- Add a binary flag (is_processorA) to identify transactions processed by ProcessorA.
- One-hot encode payment_status.
- Add a binary flag (is_successful) for transactions marked as successful.
- Create a feature (customer_name_count) to record the number of occurrences of a customer name for the merchant.
- Add a binary flag (is_suspicious_name) for names appearing frequently in fraud transactions.
- Despite the lack of statistical significance, retain payment_completion_time_secs as a numerical feature. It might contribute in combination with other features.

  
 <b> COMBINATIONS </b>
- merchant_bears_cost X merchant_risk_level
- payment_description X merchant
- txn_day_of_week X txn_hour 

In [291]:
columns_to_drop = ['merchant_bears_cost', 'merchant_bears_cost.1']
df1 = df1.drop(columns_to_drop, axis=1)

In [293]:
df1.head()

wallet_id     payment_description  payment_created_on payment_completed_on  \
0        191        Deposit 12822619 2024-05-22 16:51:06  2024-05-22 17:00:05   
1        191        Deposit 11197633 2023-05-09 18:45:47  2023-05-09 18:48:13   
2        191            FXTM 1311457 2022-09-23 11:52:21  2022-09-23 11:54:04   
3        191  Payment on Exinity Ltd 2023-07-25 04:18:40  2023-07-25 04:19:31   
4        191        Deposit 11054290 2023-04-11 05:37:20  2023-04-11 05:40:01   

   for_disbursement_wallet allowed_payment_methods     merchant  \
0                        0                      bt  Exinity Ltd   
1                        0                      bt  Exinity Ltd   
2                        0                 bt_card  Exinity Ltd   
3                        0                    card  Exinity Ltd   
4                        0                      bt  Exinity Ltd   

    account_created_on  account_risk_level payment_source_type  \
0  2021-08-18 12:15:35  above_average_risk       bank_transfer   
1  2021-08-18 12:15:35  above_average_risk       bank_transfer   
2  2021-08-18 12:15:35  above_average_risk       bank_transfer   
3  2021-08-18 12:15:35  above_average_risk                card   
4  2021-08-18 12:15:35  above_average_risk       bank_transfer   

         processor payment_status payment_channel  amount_collected  \
0             wema        success           modal           40000.0   
1         sterling        success           modal           13000.0   
2             wema        success           modal           20000.0   
3  mastercard-wema        success           modal            8000.0   
4         sterling        success           modal           70000.0   

   payment_reversals_type            customer_name  is_fraud  \
0                       0  Abubakar Ibrahim Sunusi         0   
1                       0               Mudi Aliyu         0   
2                       0       Yusuf Yahaya aliyu         0   
3                       0           Ayomide Bolaji         0   
4                       0      Richard Idris Usman         0   

   payment_completion_time_secs  merchant_tenure_in_months created_dow  \
0                         539.0                         39   Wednesday   
1                         146.0                         39     Tuesday   
2                         103.0                         39      Friday   
3                          51.0                         39     Tuesday   
4                         161.0                         39     Tuesday   

   created_hod created_month completed_dow  completed_hod completed_month  \
0           16           May     Wednesday             17             May   
1           18           May       Tuesday             18             May   
2           11     September        Friday             11       September   
3            4          July       Tuesday              4            July   
4            5         April       Tuesday              5           April   

  created_date  allowed_payment_methods_count tenure_bin  
0   2024-05-22                              1      39-42  
1   2023-05-09                              1      39-42  
2   2022-09-23                              2      39-42  
3   2023-07-25                              1      39-42  
4   2023-04-11                              1      39-42

In [295]:
df1.shape

(2604171, 28)

In [300]:
df1.to_csv('data/fraud_eda_data.csv', index=False)